src/wifi.rs

```diff
@@ -319,6 +325,57 @@ static mut RX_CALLBACK: Option<
 #[allow(clippy::type_complexity)]
 static mut TX_CALLBACK: Option<Box<dyn FnMut(WifiDeviceId, &[u8], bool) + 'static>> = None;
 
+pub trait NonBlocking {
+    fn is_scan_done(&self) -> Result<bool, EspError>;
+
+    fn start_scan(
+        &mut self,
+        scan_config: &config::ScanConfig,
+        blocking: bool,
```

Comment:
why does this fn still take an blocking bool ? Wouldn't it be nicer to create an WifiDriver in an blocking or nonblocking state and depending on that it will be set true in the blocking impl and false in the non blokcking? Or is it to not break the api to hard?

Reply:
The base driver is in fact nonblocking (event loop based) but not async. It implements the `Wifi` trait because, well, the trait actually does not specify if its methods block or not. What you are asking for is now implemented in the `AsyncWifi` and `BlockingWifi` wrappers, which are just wrapping the base nonblocking driver and wait on the event loop either asynchronously or in a blocking way. And *also* implement either the `Wifi` or the `WifiAsync` trait.

The actual problem here is that the `NonBlocking` trait is marked as `pub` but it is not an API - it is just an internal implementation trick that allows AsyncWifi and BlockingWifi to wrap both the nw level 2  driver, as well as the nw level 3 `EspWifi` thing. I'll see if I can hide this trait so that folks are not confused.

Reply:
Yeah. not great, but I can't make `NonBlocking` non-`pub` without bloating the existing implementation even further. And I just bloated it a lot - in that - besides `BlockingWifi` and `AsyncWifi`, we now also have `BlockingEth`, `AsyncEth`, `BlockingNetif`, `AsyncNetif` and `NetifStatus` (the latter we probably need to promote to `embedded-svc` at some point when we believe we have a good, trait-able abstraction for a Network Interface.